ocaml/xapi/certificates_sync.ml

```diff
@@ -0,0 +1,144 @@
+        let time = Ptime_clock.now () in
+        let tls_on = is_tls_verification_enabled () in
+        match
+          GV.validating_trust_anchors ~pem_leaf ~pkcs8_private_key
+            ~pem_intermediates ~time ~ca_cert_bundle_path
+        with
+        | Error _ as error_msg when tls_on ->
+            error "Host certificate is invalid but required by TLS" ;
+            error_msg
+        | Ok (host_cert, roots) ->
+            let hashes = List.map X509.Certificate.encode_pem roots in
+            let _ref =
+              Xapi_host.add_certificate_to_db ~__context ~host host_cert hashes
+            in
+            R.ok ()
+        | Error _ as error_msg ->
+            warn "The host certificate is invalid but TLS is off" ;
+            error_msg
+      )
+  with e -> Error (`Msg ("installation of host certificate failed", []))
+
+(** determine if the database is up to date by comparing the fingerprint
+  of xapi-ssl.pem with the entry in the database *)
+let is_unchanged ~__context cert_ref cert =
+  let host_pem = cert.GP.host_cert in
+  let ref_hash = Db.Certificate.get_fingerprint ~__context ~self:cert_ref in
+  let* host_cert =
+    Cstruct.of_string host_pem
+    |> X509.Certificate.decode_pem
+    |> R.reword_error (fun (`Msg msg) ->
+           D.info {|Failed to decode certificate because "%s"|} msg ;
+           `Msg (server_certificate_invalid, []))
+  in
+  let cert_hash =
+    X509.Certificate.fingerprint Mirage_crypto.Hash.(`SHA256) host_cert
+    |> Certificates.pp_hash
+  in
+  Ok (cert_hash = ref_hash)
+
+(** [get_server_cert] loads xapi-ssl.pem from the file system and
+  returns it as a record containing the PEM strings *)
+let get_server_cert () =
+  match G.Pem.parse_file !Xapi_globs.server_cert_path with
+  | Error msg ->
+      Error (`Msg (msg, []))
+  | Ok _ as ok ->
+      ok
+
+let update ~__context =
+  let host = Helpers.get_localhost ~__context in
+  let host_uuid = Helpers.get_localhost_uuid () in
+  let cert_refs = Db.Host.get_certificates ~__context ~self:host in
+  let* cert = get_server_cert () in
+  match cert_refs with
+  | [] ->
+      info "Host %s has no active server certificate" host_uuid ;
+      install ~__context ~host cert
+  | [cert_ref] ->
+      let* unchanged = is_unchanged ~__context cert_ref cert in
+      if unchanged then (
+        info "Active server certificate for host %s is unchanged" host_uuid ;
+        Ok ()
+      ) else (
+        info "Server certificate for host %s changed - updating" host_uuid ;
+        let* () = install ~__context ~host cert in
+        remove_cert ~__context cert_ref
+      )
+  | cert_refs ->
+      info "Server certificate for host %s changed - updating" host_uuid ;
+      let* () = install ~__context ~host cert in
+      let rec loop = function
+        | [] ->
+            Ok ()
+        | c :: cs ->
+            remove_cert ~__context c >>= fun () -> loop cs
+      in
+      loop cert_refs
```

Comment:
This can be done with:
```ocaml
      List.iter (remove_cert  ~__context) cert_refs
```
By changing remove_cert to return `()`.

This makes it more obvious that the code works with the empty case as well, if the error message is massaged a bit. I'm thinking along the lines of `info "Server certificate in the database is not the same as in the filesystem, removing the existing %i certificates for host %s" (List.length cert_refs) host_uuid;`

Reply:
There is a difference: My loop fails with an error when `remove_cert` returns an error whereas using `List.iter` would continue and gobble up the error. Not sure what the right approach is here.

Reply:
Doesn't remove_cert return always `Ok ()`?

Reply:
You are right - so this is purely hypothetical and could be simplified. Will do.